src/mcts/search.cc

```diff
@@ -135,56 +144,72 @@ void ApplyDirichletNoise(Node* node, float eps, double alpha) {
 }
 } // namespace
 
-// Returns a child with most visits.
-Node* Search::GetBestChild(Node* parent) const {
-  Node* best_node = nullptr;
-  // Best child is selected using the following criteria:
-  // * Largest number of playouts.
-  // * If two nodes have equal number:
-  //   * If that number is 0, the one with larger prior wins.
-  //   * If that number is larger than 0, the one wil larger eval wins.
-  std::tuple<int, float, float> best(-1, 0.0, 0.0);
-  for (Node* node : parent->Children()) {
-    if (parent == root_node_ && !limits_.searchmoves.empty() &&
-        std::find(limits_.searchmoves.begin(), limits_.searchmoves.end(),
-                  node->GetMove()) == limits_.searchmoves.end()) {
-      continue;
-    }
-    std::tuple<int, float, float> val(node->GetN(), node->GetQ(-10.0, 0.0),
-                                      node->GetP());
-    if (val > best) {
-      best = val;
-      best_node = node;
+float Search::CalculateTemperature() const {
+  // Put a floor on temperature to limit powers computed. Temps below the floor
+  // are rounded to 0. We round each component separately for better control.
+  float floor = 0.10f; // semi-arbitrary
+
+  // movetemp ~ 1/log(move*decay)
+  int move = played_history_.Last().GetGamePly() / 2;
+  float movetemp = kMoveTemp / (1 + std::log(1 + kMoveTempDecay * move));
+  if (movetemp < floor) movetemp = 0.0f;
+
+  // evaltemp ~ 1/log(eval^4*decay)
+  float eval = -root_node_->GetQ(0, 0);
+  float evaltemp = kEvalTemp /
+                   (1 + std::log(1 + kEvalTempDecay * std::pow(eval, 4)));
+  if (evaltemp < floor) evaltemp = 0.0f;
+
+  float temperature = movetemp + evaltemp;
+  //std::cout << "move: " << move << " movetemp: " << movetemp << " eval: " << eval << " evaltemp: " << evaltemp << " temp: " << temperature << "\n";
+
+  return temperature;
+}
+
+// Returns a child with most visits, if temp==0, or child chosen randomly
+// according to (visits_fraction)^(1/temp)
+Node* Search::GetBestChild(Node* parent,
+                           float temperature) const {
+
+  if (temperature <= 0.0f || parent->GetN() <= 1) {
+    // temp shouldn't be < 0 ever, but whatever
```

Comment:
Capitalize first word, use full word "temperature" unless it would cause the comment to break into multiple lines, period in the end.